src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs

```diff
@@ -29309,30 +29325,275 @@ class C<T>
+                );
+        }
+
+        [Theory]
+        [InlineData("T")]
+        [InlineData("TClass")]
+        [InlineData("TNotNull")]
+        [WorkItem(39922, "https://github.com/dotnet/roslyn/issues/39922")]
+        public void EnforcedInMethodBody_MaybeNullWhen_MaybeDefaultValue(string type)
         {
             var source = @"
 #nullable enable
 
 using System.Diagnostics.CodeAnalysis;
 
-class Program
+class C<T, TClass, TNotNull>
+    where TClass : class
+    where TNotNull : notnull
 {
-    static void GetValue<T>([AllowNull]T x, [MaybeNull]out T y)
+    static void GetValue([AllowNull] TYPE x, [MaybeNull] out TYPE y)
+    {
+        y = x;
+    }
+
+    static bool TryGetValue([AllowNull] TYPE x, [MaybeNullWhen(true)] out TYPE y)
+    {
+        y = x;
+        return y == null;
+    }
+
+    static bool TryGetValue2([AllowNull] TYPE x, [MaybeNullWhen(true)] out TYPE y)
+    {
+        y = x;
+        return y != null; // 1
+    }
+
+    static bool TryGetValue3([AllowNull] TYPE x, [MaybeNullWhen(false)] out TYPE y)
     {
         y = x;
+        return y == null; // 2
     }
 
-    static bool TryGetValue<T>([AllowNull]T x, [MaybeNullWhen(false)]out T y)
+    static bool TryGetValue4([AllowNull] TYPE x, [MaybeNullWhen(false)] out TYPE y)
     {
         y = x;
         return y != null;
     }
 }
 ";
-            var comp = CreateNullableCompilation(new[] { AllowNullAttributeDefinition, MaybeNullAttributeDefinition, MaybeNullWhenAttributeDefinition, source });
-            comp.VerifyDiagnostics();
+            var comp = CreateNullableCompilation(new[] { AllowNullAttributeDefinition, MaybeNullAttributeDefinition, MaybeNullWhenAttributeDefinition, source.Replace("TYPE", type) });
+            comp.VerifyDiagnostics(
+                // (24,9): error CS8762: Parameter 'y' may not have a null value when exiting with 'false'.
+                //         return y != null; // 1
+                Diagnostic(ErrorCode.WRN_ParameterConditionallyDisallowsNull, "return y != null;").WithArguments("y", "false").WithLocation(24, 9),
+                // (30,9): error CS8762: Parameter 'y' may not have a null value when exiting with 'true'.
+                //         return y == null; // 2
+                Diagnostic(ErrorCode.WRN_ParameterConditionallyDisallowsNull, "return y == null;").WithArguments("y", "true").WithLocation(30, 9)
+                );
+        }
+
+
+        [Fact, WorkItem(39922, "https://github.com/dotnet/roslyn/issues/39922")]
+        public void EnforcedInMethodBody_MaybeNullWhen_TwoParameter()
+        {
+            var source = @"
+#nullable enable
+
+using System.Diagnostics.CodeAnalysis;
+
+class Program
+{
+    static bool TryGetValue<T>([AllowNull]T x, [MaybeNullWhen(true)]out T y, [MaybeNullWhen(true)]out T z)
+    {
+        y = x;
+        z = x;
+        return y != null || z != null;
```

Comment:
What does the "happy case" for this look like? Perhaps `y == null || z == null`? Is that test present here?

Reply:
➡️ added test